### PR TITLE
ICU-22261 Add tech preview implementation for MessageFormat 2.0 to icu4c

### PR DESCRIPTION
**Pull request by catamorphism**
This pull request implements the [draft MessageFormat 2.0 spec](https://github.com/unicode-org/message-format-wg); the goal is to implement the same functionality as the existing ICU4J tech preview for MessageFormat 2.0, though there are differences since the code in this pull request reflects a newer version of the spec.

[A design document](https://docs.google.com/document/d/1qi5zeYevPshu-Gr8-nn7Rsq-WFOAx_zNxOVJkRz5cWM/edit?resourcekey=0-dUYbqvZh5vTEJRPNoZ4ehA) has been shared with the icu-design mailing list, with the deadline for commenting as 2023-09-05, though this may be extended if not enough feedback is received. There are some open comments on the design document that I'll address, but none of them seem to be a blocker for initial review of this draft pull request.

##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22261
- [X] Required: The PR title must be prefixed with a JIRA Issue number. 
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [x] Issue accepted (done by Technical Committee after discussion)
- [X] Tests included, if applicable
- [X] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

**Comment by echeran** [comment]
@catamorphism In the ICU-TC meeting, @richgillam mentioned that when he looked last time, which was a few months ago, there was a discrepancy in functionality covered between our existing ICU4J Tech Preview implementation and what the ICU4C MF2 PR covered. I think you and I discussed (also a while ago) that having a first step PR to reach parity with the ICU4J code would be easier for everyone (breaks up work into chunks, easier for reviewers, makes API proposal review much simpler). 

And it looks like your API proposal doc is doing that, from my very quick superficial inspection, which is great.  (Although while I'm at it, double  check that the signature `void getPattern(UnicodeString&) const;` makes sense.) You can also port the unit tests from the ICU4J, which hopefully is easier for you, and it definitely helps maintainers keep the 2 languages' impls in sync. It looks like you're making updates on the PR, which is cool, although I haven't looked at it yet. Thanks for your patience and ongoing work.

**Comment by echeran** [comment]
FYI, as you work on the PR, you can keep adding commits until the very end, rather than having to squash. It also makes it easier for the reviewer, too, since the previous comments don't disappear. (Comments disappear when you force push.) We just recently pulled our guidelines into [markdown in the User Guide](https://unicode-org.github.io/icu/userguide/dev/gitdev.html#review--commit-to-unicode-main). Were you rebasing because of the "Single commit" PR checker bot CI check? Maybe we can rename that if it was confusing?

**Comment by catamorphism** [comment]
> @catamorphism In the ICU-TC meeting, @richgillam mentioned that when he looked last time, which was a few months ago, there was a discrepancy in functionality covered between our existing ICU4J Tech Preview implementation and what the ICU4C MF2 PR covered. I think you and I discussed (also a while ago) that having a first step PR to reach parity with the ICU4J code would be easier for everyone (breaks up work into chunks, easier for reviewers, makes API proposal review much simpler).

Can you (or Rich) be more specific? In writing the design document, I tried to stay very close to what was described in the ICU4J design document (and this guided my implementation too). I think the API itself in my ICU4C design doc doesn't add any extra functionality. The main difference is that the implementation in this PR reflects a more recent version of the MessageFormat syntax than the ICU4J implementation does (the latter reflects the syntax as of September 2022). That necessitates some changes to the data model and therefore to the data model API. But, I had thought the goal was for the ICU4J implementation to be updated to reflect the newer version of the syntax, rather than for me to roll back the ICU4C implementation to use that version.

**Comment by markusicu** [comment]
We had asked @mihnita to do the functional review, and talked about @roubert helping with C++.
@echeran FYI

**Comment by markusicu** [comment]
@catamorphism this PR is not really still merely a draft, right?

**Comment by catamorphism** [comment]
> @catamorphism this PR is not really still merely a draft, right?

No, not really a draft. There's a little bit more to do on the built-in functions, but that shouldn't block the review.

**Comment by roubert** [comment]
> and talked about @roubert helping with C++.

Sure thing, but that's going to need some substantial amount of time. I took a brief look 10 minutes ago (before the currently latest commit was added) and then the 34 newly added source files in this PR contained 10,452 lines of C++ code plus 5,259 lines of comments.

@catamorphism, I see that you're constantly adding new commits, please tell me once the PR is stable enough for it to make sense for me to start writing review comments.

**Comment by catamorphism** [comment]
> Sure thing, but that's going to need some substantial amount of time. I took a brief look 10 minutes ago (before the currently latest commit was added) and then the 34 newly added source files in this PR contained 10,452 lines of C++ code plus 5,259 lines of comments.
> 

Based on today's TC meeting, I believe that the TC is okay with a cursory review before the upcoming ICU release, because this is a tech preview implementation. How cursory is "cursory" is not something I know how to specify :)

> @catamorphism, I see that you're constantly adding new commits, please tell me once the PR is stable enough for it to make sense for me to start writing review comments (about the C++ implementation).

It's fine to start writing review comments now. If this was going to be a full review, we could talk about places to start, but I'm not sure what's most useful to review in a limited amount of time. @markusicu do you have any thoughts?

**Comment by catamorphism** [comment]
Just wanted to add a note about the remaining CI failures ([this](https://github.com/unicode-org/icu/actions/runs/8382036498/job/22955669770) is the most recent run of the full "GHA ICU4C" workflow -- I don't know how to find it directly from the PR, since it doesn't seem to run every time):

* `gcc-debug-build-and-test` fails with this error:

```
2024-03-21T22:06:58.5150214Z internaltest unicode/gregocal.h
2024-03-21T22:06:58.7272152Z In file included from ht_stub_internaltest.cpp:1:
2024-03-21T22:06:58.7277646Z /tmp/icu-prefix/include/unicode/gregocal.h:725:5: error: ‘DECLARE_OVERRIDE_SYSTEM_DEFAULT_CENTURY’ does not name a type
2024-03-21T22:06:58.7338816Z   725 |     DECLARE_OVERRIDE_SYSTEM_DEFAULT_CENTURY
2024-03-21T22:06:58.7339496Z       |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I'm unable to reproduce this locally, after running:

```
$ ../icu4c/source/runConfigureICU --enable-debug --disable-release Linux/gcc --prefix=/tmp/icu-prefix --enable-tracing
$ make install
$  PATH=/tmp/icu-prefix:$PATH make -C test/hdrtest check
```

I'm also not sure why it's erroring on a file I haven't changed. One of the test files does include `gregocal.h`, but I can't find any references to `DECLARE_OVERRIDE_SYSTEM_DEFAULT_CENTURY` anywhere in the codebase. I know the error is coming from preprocessed code, but that's as far as I got.

* The other failures are in the `source/test/hdrtst/testtagsguards.sh` script. I can repro that one locally, but I'm having difficulty figuring out what the script is doing or what the error is (since the output is 200K lines of preprocessed code with no explanatory error message.)

**Comment by richgillam** [comment]
Tim—

The SYSTEM_DEFAULT_CENTURY thing was a mistake Frank checked in my accident a day or two ago.  If you rebase your PR on top of the current ToT in main, that ought to go away (or at least it did for me on one of my PRs).

—Rich

**Comment by catamorphism** [comment]
Thanks @richgillam ! I rebased (which required me to force-push -- hopefully that won't cause problems, since there are no line comments yet to lose).

**Comment by catamorphism** [comment]
>     * The other failures are in the `source/test/hdrtst/testtagsguards.sh` script. I can repro that one locally, but I'm having difficulty figuring out what the script is doing or what the error is (since the output is 200K lines of preprocessed code with no explanatory error message.)

Figured it out and fixed it in [b07d4b5](https://github.com/unicode-org/icu/pull/2579/commits/b07d4b5b08fd0f70c434295f29ab9e8fdc212335).

**Comment by markusicu** [comment]
> If this was going to be a full review, we could talk about places to start, but I'm not sure what's most useful to review in a limited amount of time. @markusicu do you have any thoughts?

We do want to get this into 75 as a tech preview, and we do want to tag the 75 branch asap, so that we can work on getting the release candidate out of the door.

So for review, I know this is large, the most important part right now is whether the public API looks ok with spelling, visibility, constness and such. Maybe even try to look at the public parts next to the design doc. And maybe @roubert could take a look at what types cross the DLL boundary, modulo the discussion we had (and documented in the design doc) about std::variant generating exception-throwing code even though the MF2 code does not call those functions.

@echeran and maybe @mihnita should probably look over the test code to see if anything looks off.

For the bulk of the code, not sure what to say. Large PRs are hard. We probably need to rubber-stamp a lot of it for now. We can still fix some bugs between release candidate and final (in follow-up PRs), but we should not change the API between rc and final.

**Comment by catamorphism** [comment]
> hopefully that won't cause problems

Sorry, looks like force-pushing did cause problems (I see that some of the comments have disappeared even though they were not line comments). I'll try to avoid doing that again.

**Comment by markusicu** [comment]
The good news is that all of the CI checks are passing, except for the two that will keep complaining until you squash your commits -- but don't squash until the reviewers have approved.

File: icu4c/source/common/unicode/utypes.h
Comment (catamorphism) [reply icu4c/source/common/unicode/utypes.h:574]: It looks like some of the other APIs do (`U_BRK_*` for BreakIterator, `U_REGEX_*` for Regex, etc.) I'm fine with that, but will wait for feedback about what the prefix should be.

File: icu4c/source/i18n/messageformat2_allocation.h
Comment (catamorphism) [reply icu4c/source/i18n/messageformat2_allocation.h]: Fixed

File: icu4c/source/i18n/messageformat2_arguments.cpp
Comment (catamorphism) [reply icu4c/source/i18n/messageformat2_arguments.cpp]: I was depending on assertions for that, but you're right, it's not a great pattern. Changed in [aa828b3](https://github.com/unicode-org/icu/pull/2579/commits/aa828b3ab0ffcce840ac4d8fc4812600c995e880) to use error codes instead.

File: icu4c/source/i18n/messageformat2_checker.h
Comment (catamorphism) [reply icu4c/source/i18n/messageformat2_checker.h]: Good point. Changed this in [1184d7c](https://github.com/unicode-org/icu/pull/2579/commits/1184d7cc99e41e8abceb1ad4f0929112d32f78bb) and added a comment -- let me know if that makes sense.

File: icu4c/source/i18n/messageformat2_errors.h
Comment (catamorphism) [reply icu4c/source/i18n/messageformat2_errors.h:148]: Good point. I got there by copy/pasting. I just looked at some existing public header files, and they are similarly inconsistent (for example, I looked at numberformatter.h, listformatter.h, formattednumber.h). Since existing code doesn't consistently use one or the other, I would rather keep it as it is, since I don't know which way is correct.

File: icu4c/source/i18n/messageformat2_formatter.cpp
Comment (catamorphism) [reply icu4c/source/i18n/messageformat2_formatter.cpp]: Fixed

File: icu4c/source/i18n/messageformat2_formatter.cpp
Comment (catamorphism) [reply icu4c/source/i18n/messageformat2_formatter.cpp]: Fixed

File: icu4c/source/test/intltest/messageformat2test_builtin.cpp
Comment (catamorphism) [reply icu4c/source/test/intltest/messageformat2test_builtin.cpp]: Good point -- this is an actual bug. According to the spec, it should be an error ( https://github.com/unicode-org/message-format-wg/blob/main/spec/registry.md#number-operands ). I wrote this code before that section of the spec existed, and didn't notice it was wrong because there weren't spec tests that would contradict what the expected output was here.

I fixed this in [cecdf30](https://github.com/unicode-org/icu/pull/2579/commits/cecdf30f32fd661e64cb2a6931562579af794a59) (and noted in comments there that it's not the best solution, but it's what I had time for.)

File: icu4c/source/common/unicode/utypes.h
Comment (markusicu) [reply icu4c/source/common/unicode/utypes.h:574]: good idea; U_MF_ seems fine

Comment (catamorphism) [reply icu4c/source/common/unicode/utypes.h]: Done in [0dc45b0](https://github.com/unicode-org/icu/pull/2579/commits/0dc45b077905548a3a2ddbb4249d0a4eee8bfc34)

**Comment by markusicu** [comment]
Can y'all please resolve the remaining comments -- looks like Tim took care of them as much as possible for now -- and move to getting this merged by tomorrow=Wednesday noon Pacific Time?

Some bug fixes can go into the maintenance branch, and more changes can go into ICU 76.

**Comment by catamorphism** [comment]
As far as I could tell there was nothing actionable in the remaining open comments, so I resolved them, but feel free to reopen @mihnita if there's anything in there that you still think I need to address.

**Review by mihnita — DISMISSED** [review]
Thank you!
M.

**Review by macchiati — APPROVED** [review]

**Review by roubert — COMMENTED** [review]
If I should offer a very high-level opinion on this, it's that it might be a very good idea to in addition to the already existing `UCONFIG_NO_FORMATTING` flag also add one such flag specifically for this new MF2 functionality, to make it easy to build ICU4C without any of it, for those who won't be using this tech preview and would prefer to not have tens of thousands of lines of new mostly unreviewed code included in their binaries.

**Comment by markusicu** [comment]
> If I should offer a very high-level opinion on this, it's that it might be a very good idea to in addition to the already existing `UCONFIG_NO_FORMATTING` flag also add one such flag specifically for this new MF2 functionality, to make it easy to build ICU4C without any of it, for those who won't be using this tech preview and would prefer to not have tens of thousands of lines of new mostly unreviewed code included in their binaries.

Good idea!
Let's add that in a follow-up PR, possibly on the maint/maint-75 branch.
Something like `UCONFIG_NO_MF2` ?

**Comment by catamorphism** [comment]
> Good idea! Let's add that in a follow-up PR, possibly on the maint/maint-75 branch. Something like `UCONFIG_NO_MF2` ?

I would be happy to do that in a follow-up PR. Note that as the code author, I am just as nervous about shipping unreviewed code as users might be about including that code in their binaries :)

**Comment by echeran** [comment]
> @echeran ok to squash and then merge?

Yep, let's squash & merge now.

**Review by echeran — APPROVED** [review]
RSLGTM

**Comment by echeran** [comment]
(note: For anyone noticing the PR checker bot claiming that the branch contents changed across the force push: I don't think that is the case. I did the squash locally since I cannot use the PR checker bot to squash on another person's PR branch (PR checker bot does not have permissions to do so). When I run `git diff for-review..catamorphism/for-review` locally, I get no output and a 0 return code. Visual inspection of the file, which is an update to the `Makefile.in` for tests, seems correct and unchanged.)

**Comment by markusicu** [comment]
Thanks everyone!! 🎉

**Comment by markusicu** [comment]
> > Good idea! Let's add that in a follow-up PR, possibly on the maint/maint-75 branch. Something like `UCONFIG_NO_MF2` ?
> 
> I would be happy to do that in a follow-up PR.

Please do now :-)

**Comment by catamorphism** [comment]
> > > Good idea! Let's add that in a follow-up PR, possibly on the maint/maint-75 branch. Something like `UCONFIG_NO_MF2` ?
> > 
> > 
> > I would be happy to do that in a follow-up PR.
> 
> Please do now :-)

Submitted: #2931

**Pull request merged**
